src/transformers/models/phimoe/modeling_phimoe.py

```diff
@@ -0,0 +1,1850 @@
+        self.register_buffer("inv_freq", inv_freq, persistent=False)
+
+        # Build here to make `torch.jit.trace` work.
+        self._set_cos_sin_cache(
+            seq_len=max_position_embeddings, device=self.inv_freq.device, dtype=torch.get_default_dtype()
+        )
+
+    def _set_cos_sin_cache(self, seq_len, device, dtype):
+        self.max_seq_len_cached = seq_len
+        t = torch.arange(self.max_seq_len_cached, device=device, dtype=torch.int64).type_as(self.inv_freq)
+
+        freqs = torch.outer(t, self.inv_freq)
+        # Different from paper, but it uses a different permutation in order to obtain the same calculation
+        emb = torch.cat((freqs, freqs), dim=-1)
+        self.register_buffer("cos_cached", emb.cos().to(dtype), persistent=False)
+        self.register_buffer("sin_cached", emb.sin().to(dtype), persistent=False)
+
+    def forward(self, x, seq_len=None):
+        # x: [bs, num_attention_heads, seq_len, head_size]
+        if seq_len > self.max_seq_len_cached:
+            self._set_cos_sin_cache(seq_len=seq_len, device=x.device, dtype=x.dtype)
+
+        return (
+            self.cos_cached[:seq_len].to(dtype=x.dtype),
+            self.sin_cached[:seq_len].to(dtype=x.dtype),
+        )
+
+
+class Phi3LongRoPEScaledRotaryEmbedding(nn.Module):
+    def __init__(self, dim, config):
+        super().__init__()
+        self.dim = dim
+        self.max_position_embeddings = config.max_position_embeddings
+        self.base = config.rope_theta
+        self.short_factor = config.rope_scaling["short_factor"]
+        self.long_factor = config.rope_scaling["long_factor"]
+        self.short_mscale = config.rope_scaling["short_mscale"]
+        self.long_mscale = config.rope_scaling["long_mscale"]
+        self.original_max_position_embeddings = config.rope_scaling["original_max_position_embeddings"]
+
+    def forward(self, x, seq_len=None):
+        if seq_len is None:
+            seq_len = x.shape[-2]
+
+        if seq_len > self.original_max_position_embeddings:
+            rescale_factors = torch.tensor(self.long_factor, dtype=torch.float32, device=x.device)
+            mscale = self.long_mscale
+        else:
+            rescale_factors = torch.tensor(self.short_factor, dtype=torch.float32, device=x.device)
+            mscale = self.short_mscale
+        assert rescale_factors.shape == (
+            self.dim // 2,
+        ), f"misaligned shape for LongRoPE rescale factors: {rescale_factors.shape}"
+
+        inv_freq = 1.0 / (
+            rescale_factors * (self.base ** (torch.arange(0, self.dim, 2).float().to(x.device) / self.dim))
+        )
+
+        t = torch.arange(seq_len, device=x.device, dtype=torch.float32)
+        freqs = torch.outer(t, inv_freq)
+
+        emb = torch.cat((freqs, freqs), dim=-1)
+        return (emb.cos() * mscale).to(x.dtype), (emb.sin() * mscale).to(x.dtype)
+
```

Comment:
```suggestion
```
new rope scaling should be added to this: https://github.com/huggingface/transformers/blob/7e638ef2b8650aaa3e3a8e575bb63af262a43d95/src/transformers/modeling_rope_utils.py#L353-L360

Reply:
Modified the longrope implementation to use short factors for shorter sequences when seq_len is less than original_max_position_embeddings.